### PR TITLE
Update close.io for changes in tap-tester

### PR DESCRIPTION
**Pull request by HarrisonMarcRose**
# Description of change
tests within the repo are sometimes hard to follow.  This will attempt to make it clearer where things are coming from 

make start_date an instance property - not a class attribute, be specific about class methods instead of self, implement smaller test override

# Manual QA steps
 - 
 
# Risks
 - 
 
# Rollback steps
 - revert this branch

### CODE REVIEW

File: tests/test_bookmark.py [hunk 1]
```diff
@@ -1,39 +1,60 @@
...
-    def __init__(self, test_run):
-        super().__init__(test_run)
-
-        #Hard coded bookmark_date as there are no active records
-        self.bookmark_date = '2018-01-25T00:00:00Z'
+        # TODO - are there no records for the event_log stream?
+        # return {'tasks'}
+        return self.expected_stream_names().difference({'event_log'})
+
+    def test_first_sync_bookmark(self):
```
Comment (bhtowles) [inline tests/test_bookmark.py:26]: Can we discuss what these two methods are doing?

File: tests/test_all_fields.py [hunk 2]
```diff
@@ -5,13 +5,103 @@
 from tap_tester.base_suite_tests.all_fields_test import AllFieldsTest
 from base import CloseioBase
 
-class CloseioAllFieldsTest(AllFieldsTest,CloseioBase):
+
+class CloseioAllFieldsTest(AllFieldsTest, CloseioBase):
     """Test that with no fields selected for a stream automatic fields are still replicated"""
 
+    selected_fields = AllFieldsTest.selected_fields
```
Comment (bryantgray) [inline tests/test_all_fields.py:12]: Should this be `CloseioAllFieldsTests.selected_fields`? It probably doesn't actually makes a difference, but it reads like we're taking `selected_fields` that are predefined in `AllFieldsTest`. Since `AllFieldsTest` is an abstract class, it feels a little odd to use the `selected_fields` property from it instead of the implementation of that class (`CloseioAllFieldsTests`).

File: tests/test_all_fields.py [hunk 3]
```diff
@@ -5,13 +5,103 @@
...
+                    'subject',
+                    'voicemail_duration',
+                    'voicemail_url',
+                }}
+            additional_expected = {
+                stream: fields.union(
+                    extra_fields.get(stream, set())).difference(
+                    missing_fields.get(stream, set()))
+                for stream, fields in actual_expected.items()}
+            return additional_expected
```
Comment (bryantgray) [inline tests/test_all_fields.py:105]: Looks like this can be moved to base.py since it is used for test_automatic_fields as well.

File: tests/test_bookmark.py [hunk 4]
```diff
@@ -1,39 +1,60 @@
...
+
 class CloseioBookmarkTest(BookmarkTest, CloseioBase):
     """Closeio bookmark test implementation"""
 
+    bookmark_format = "%Y-%m-%dT%H:%M:%S.%f+00:00"
+    initial_bookmarks = {
+            'bookmarks': {
+                'custom_fields': {'date_updated': '2016-07-07T14:22:04.624000Z'},
+                'leads': {'date_updated': '2020-05-06T18:54:08.250000Z'},
+                'activities': {'date_created': '2020-11-06T20:17:17.454000Z'},
```
Comment (bryantgray) [inline tests/test_bookmark.py:13]: Line 59 mentions the `activities` stream having a different bookmark format that is missing the fractional seconds. Should we change it to match that format here?

**Review by bryantgray — COMMENTED** [review]
Just a few small comments. I did notice `BaseCase.API_LIMIT` is actually the page limit. This wasn't super clear at first and my mind immediately went to API quotas/rate-limiting. This is a tricky one though because I'm not sure naming it `PAGE_LIMIT` in the base suite would always be accurate. I'm interested to hear thoughts on this.

File: tests/test_all_fields.py (hunk 2 above)
Comment (HarrisonMarcRose) [reply tests/test_all_fields.py:12]: The selected_fields is defined during setup.  The only reason this line is here is pylint.  This line doesn't actually perform any function except to let the rest of the class know where this is defined.

I could just remove this line.  Is that what you are suggesting?

The abstract classes do a combination of implementing methods that it can and defining which ones need to be defined.

**Review by bhtowles — APPROVED** [review]

File: tests/test_discovery.py [hunk 5]
```diff
@@ -7,49 +6,36 @@
 class CloseioDiscoveryTest(DiscoveryTest, CloseioBase):
     """Standard Discovery Test"""
 
+    expected_replication_keys = CloseioBase.expected_replication_keys
```
[the remark this answers is not in the document]
Comment (HarrisonMarcRose) [reply tests/test_discovery.py:9]: This is here because pylint didn't know that we were inheriting from CloseioBase.  This is just so that pylint was happy, but functionally this doesn't do anything because the inheritance does the same thing.

File: tests/test_all_fields.py [hunk 6]
```diff
@@ -5,13 +5,103 @@
...
     def streams_to_test(self):
         streams_to_exclude = {'event_log', 'custom_fields'}
-        return set(self.expected_metadata().keys()) - streams_to_exclude
+        return self.expected_stream_names().difference(streams_to_exclude)
+
+    def test_all_fields_for_streams_are_replicated(self):
+        # TODO - BUG it looks like there is no metadata for some fields even though they
+        #   are actually replicated.  This isn't right.  Maybe it doesn't matter too much
+        #   since everything is automatic but we should write this up.
+        def too_many_fields_replicated():
```
[the remark this answers is not in the document]
Comment (HarrisonMarcRose) [reply tests/test_all_fields.py]: done
